File: java/src/main/java/com/pbdviewer/utils/pbclass/PbFunction.java
package com.pbdviewer.utils.pbclass;

public class PbFunction {

    private final PbProject project;
    private final PbEntry entry;
    private final PbObject object;
    private int index;
    private PbFunctionDefinition definition;
    private byte[] pCodeBytes;
    private byte[] debugBytes;
    private byte[] buffer;
    private PbVariable[] variables;

    public PbFunction(PbObject object) {
        this.object = object;
        this.entry = object.getEntry();
        this.project = entry.getProject();
    }

    @Override
    public String toString() {
        String defName = definition != null ? definition.getName() : String.format("#%04X", index);
        return object + "/" + defName;
    }

    public PbProject getProject() {
        return project;
    }

    public PbEntry getEntry() {
        return entry;
    }

    public PbObject getObject() {
        return object;
    }

    public int getIndex() {
        return index;
    }

    public void setIndex(int index) {
        this.index = index;
    }

    public PbFunctionDefinition getDefinition() {
        return definition;
    }

    public void setDefinition(PbFunctionDefinition definition) {
        this.definition = definition;
    }

    public byte[] getPCodeBytes() {
        return pCodeBytes;
    }

    public void setPCodeBytes(byte[] pCodeBytes) {
        this.pCodeBytes = pCodeBytes;
    }

    public byte[] getDebugBytes() {
        return debugBytes;
    }

    public void setDebugBytes(byte[] debugBytes) {
        this.debugBytes = debugBytes;
    }

    public byte[] getBuffer() {
        return buffer;
    }

    public void setBuffer(byte[] buffer) {
        this.buffer = buffer;
    }

    public PbVariable[] getVariables() {
        return variables;
    }

    public void setVariables(PbVariable[] variables) {
        this.variables = variables;
    }
}

File: html/js/pbclass/PbFunction.js
// PbFunction.js — port of Java PbFunction.java
export class PbFunction {
    /**
     * @param {import('./PbObject.js').PbObject} object
     */
    constructor(object) {
        this.object  = object;
        this.entry   = object.entry;
        this.project = object.entry.getProject();

        /** @type {number} */
        this.index       = 0;
        /** @type {import('./PbFunctionDefinition.js').PbFunctionDefinition|null} */
        this.definition  = null;
        /** @type {Uint8Array} */
        this.pCodeBytes  = new Uint8Array(0);
        /** @type {Uint8Array} */
        this.debugBytes  = new Uint8Array(0);
        /** @type {Uint8Array} */
        this.buffer      = new Uint8Array(0);
        /** @type {import('./PbVariable.js').PbVariable[]} */
        this.variables   = [];
    }

    toString() {
        const defName = this.definition !== null
            ? this.definition.name
            : `#${this.index.toString(16).padStart(4,'0').toUpperCase()}`;
        return `${this.object}/${defName}`;
    }
}
